### PR TITLE
.Net: Added AzureChatExtensionsOptions property and deprecated WithData classes

### PR DESCRIPTION
**Pull request by dmytrostruk**
### Motivation and Context

Fixes: https://github.com/microsoft/semantic-kernel/issues/4787
Fixes: https://github.com/microsoft/semantic-kernel/issues/6061
Fixes: https://github.com/microsoft/semantic-kernel/issues/6448
Fixes: https://github.com/microsoft/semantic-kernel/issues/6112

Discussion: https://github.com/microsoft/semantic-kernel/discussions/6016#discussioncomment-9459380

There are two reasons why this change is required:
1. Azure .NET SDK supports `chat completion on your data` functionality out-of-the-box, which means that now we don't need custom implementation.
2. `Chat completion on your data` new API has [breaking changes](https://learn.microsoft.com/en-us/azure/ai-services/openai/references/on-your-data), which makes current `WithData` classes non-functional. 

### Description

1. `AzureChatExtensionsOptions` property was added to `OpenAIPromptExecutionSettings`. Initial idea was to add it on `AzureOpenAIChatCompletionService` class level (in all constructors and extension methods), but it looks like this configuration contains some properties which should be useful per request (e.g. index name, search filter etc). Later, we can also add it on service class level, if needed (although I would do that when Options classes for all `AddChatCompletion...`  extension methods will be introduced, to avoid more parameters in these methods at the moment).
2. `AzureChatExtensionsOptions` property is marked with `JsonIgnore` attribute, because it's still needs to be investigated if this configuration is serializable. It supports not only Azure AI Search configuration, but also [other configurations](https://learn.microsoft.com/en-us/azure/ai-services/openai/references/on-your-data#data-source) like Azure Cosmos DB for MongoDB, Pinecone etc. We can release it without JSON serialization first, to unblock code-first customers, and we can add support for JSON serialization later (i.e. `config.json` files), if there will be such request.

Note: It looks like `OpenAIPromptExecutionSettings` is not ideal place for Azure-specific `AzureChatExtensionsOptions` property. I was thinking to add derived `AzureOpenAIPromptExecutionSettings : OpenAIPromptExecutionSettings` class (and unmark `OpenAIPromptExecutionSettings` as sealed), but this looks like some scope of work which should be implemented when OpenAI and Azure OpenAI functionality will be separated by SDKs. 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: .github/workflows/dotnet-build-and-test.yml [hunk 1]
```diff
@@ -81,7 +81,7 @@ jobs:
         run: |
           export UT_PROJECTS=$(find ./dotnet -type f -name "*.UnitTests.csproj" | grep -v -E "(Experimental.Orchestration.Flow.UnitTests.csproj|Experimental.Assistants.UnitTests.csproj)" | tr '\n' ' ')
           for project in $UT_PROJECTS; do
-            dotnet test -c ${{ matrix.configuration }} $project --no-build -v Normal --logger trx --collect:"XPlat Code Coverage" --results-directory:"TestResults/Coverage/" -- DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.ExcludeByAttribute=ObsoleteAttribute,GeneratedCodeAttribute,CompilerGeneratedAttribute,ExcludeFromCodeCoverageAttribute
+            dotnet test -c ${{ matrix.configuration }} $project --no-build -v Normal --logger trx --collect:"XPlat Code Coverage" --results-directory:"TestResults/Coverage/" -- DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.ExcludeByAttribute=GeneratedCodeAttribute,CompilerGeneratedAttribute,ExcludeFromCodeCoverageAttribute
```
Comment (markwallace-microsoft) [inline .github/workflows/dotnet-build-and-test.yml:84]: Curious why this change is needed for this PR?

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.OpenAI/OpenAIPromptExecutionSettings.cs [hunk 2]
```diff
@@ -287,6 +287,23 @@ public int ResultsPerPrompt
         }
     }
 
+    /// <summary>
+    /// An abstraction of additional settings for chat completion, see https://learn.microsoft.com/en-us/dotnet/api/azure.ai.openai.azurechatextensionsoptions.
+    /// This property is compatible only with Azure OpenAI.
+    /// </summary>
+    [Experimental("SKEXP0010")]
+    [JsonIgnore]
+    public AzureChatExtensionsOptions? AzureChatExtensionsOptions
```
Comment (SergeyMenshykh) [inline dotnet/src/Connectors/Connectors.OpenAI/OpenAIPromptExecutionSettings.cs:296]: I wonder whether an abstraction layer over the `AzureChatExtensionsOptions` class was considered as an option to isolate service API consumers from internal implementation details?

File: .github/workflows/dotnet-build-and-test.yml (hunk 1 above)
Comment (dmytrostruk) [reply .github/workflows/dotnet-build-and-test.yml:84]: `Obsolete` attribute was added to code coverage filtering in recent PRs. As I marked bunch of classes as `Obsolete`, the code coverage decreased, but I think it's incorrect. Even though we marked some classes as `Obsolete`, they are still in-place, and somebody may use it. Which means that even if we do not recommend using these classes anymore, we still need to test them and include them into code coverage, especially if these classes are located in v1 assembly (without `alpha` or `preview` suffix). That's why I removed `Obsolete` attribute from code coverage filtering. Let me know if that makes sense.

Comment (stephentoub) [reply .github/workflows/dotnet-build-and-test.yml:84]: >  As I marked bunch of classes as Obsolete, the code coverage decreased

You mean the percentage of covered lines/branches decreased? Were the classes you marked as obsolete better covered than the average? I'd have expected the WithData stuff to be, on average, less well covered, and thus I'd have expected marking them obsolete and hiding them from coverage would have actually increased percentage covered.

File: dotnet/src/Connectors/Connectors.OpenAI/OpenAIPromptExecutionSettings.cs (hunk 2 above)
Comment (dmytrostruk) [reply dotnet/src/Connectors/Connectors.OpenAI/OpenAIPromptExecutionSettings.cs:296]: Great question! I was thinking about it, but the number of possible configuration properties for this functionality is huge, here is just some of them:
image

I think it would be difficult to abstract these parameters and maintain them if anything changes in the future. I also think that abstraction could be useful if similar settings would exist in other connectors as well, but as for now it's Azure-specific settings, so I decided to expose `AzureChatExtensionsOptions` class directly, in similar way how we do that with `OpenAIClient` in Azure/OpenAI chat completion service constructors.

Let me know what you think about it.

File: .github/workflows/dotnet-build-and-test.yml (hunk 1 above)
Comment (dmytrostruk) [reply .github/workflows/dotnet-build-and-test.yml:84]: > You mean the percentage of covered lines/branches decreased?

So, without `Obsolete` attribute branch coverage for OpenAI assembly is 80.4%. With `Obsolete` attribute it's 79.5%.

> Were the classes you marked as obsolete better covered than the average?

It could be the case. Which means that we can add more tests to other OpenAI functionality to increase code coverage in more important places. But I'm still not sure if we need to exclude `Obsolete` classes from code coverage report, since deprecated functionality is still in place.

File: dotnet/src/Connectors/Connectors.OpenAI/OpenAIPromptExecutionSettings.cs (hunk 2 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.OpenAI/OpenAIPromptExecutionSettings.cs:296]: I think this solution is good for now, considering the large number of classes that require abstraction to achieve the isolation we might never need. As you mentioned, let's revise the model later when we have more connectors using similar settings.

**Review by SergeyMenshykh — APPROVED** [review]
